samples/net/mqtt/src/modules/transport/transport.c

```diff
@@ -0,0 +1,216 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/zbus/zbus.h>
+#include <zephyr/smf.h>
+#include <mqtt_helper.h>
+
+#include "client_id.h"
+#include "message_channel.h"
+
+#define MESSAGE_QUEUE_SIZE 5
+
+/* Register log module */
+LOG_MODULE_REGISTER(transport, CONFIG_MQTT_SAMPLE_TRANSPORT_LOG_LEVEL);
+
+/* Register subscriber */
+ZBUS_SUBSCRIBER_DEFINE(transport, MESSAGE_QUEUE_SIZE);
+
+/* Forward declaration of state table */
+static const struct smf_state state[];
+
+/* Internal states */
+enum module_state { MQTT_CONNECTED, MQTT_DISCONNECTED };
+
+/* User defined state object.
+ * Used to transfer data between state changes.
+ */
+static struct s_object {
+	/* This must be first */
+	struct smf_ctx ctx;
+
+	/* Last channel type that a message was received on */
+	const struct zbus_channel *chan;
+
+	/* Last channel payload that was received */
+	union {
+		/* Network status */
+		enum network_status status;
+
+		/* Payload */
+		struct payload payload;
+	};
+} s_obj;
+
+/* Callback handlers from MQTT helper library */
+static void on_connack(enum mqtt_conn_return_code return_code)
+{
+	ARG_UNUSED(return_code);
+
+	smf_set_state(SMF_CTX(&s_obj), &state[MQTT_CONNECTED]);
+}
+
+static void on_disconnect(int result)
+{
+	ARG_UNUSED(result);
+
+	smf_set_state(SMF_CTX(&s_obj), &state[MQTT_DISCONNECTED]);
+}
+
+static void on_publish(struct mqtt_helper_buf topic, struct mqtt_helper_buf payload)
+{
+	LOG_INF("Received payload: %.*s on topic: %.*s", payload.size,
+							 payload.ptr,
+							 topic.size,
+							 topic.ptr);
+}
+
+/* Zephyr State Machine framework handlers */
+static void disconnected_run(void *o)
+{
+	struct s_object *od = o;
+
+	if (od->status == NETWORK_CONNECTED) {
+		const uint8_t *client_id = client_id_get();
+
+		struct mqtt_helper_conn_params conn_params = {
+			.hostname.ptr = CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME,
+			.hostname.size = strlen(CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME),
+			.device_id.ptr = (char *)client_id,
+			.device_id.size = strlen(client_id),
+		};
+
+		int err = mqtt_helper_connect(&conn_params);
+
+		if (err) {
+			LOG_ERR("Failed connecting to MQTT, error code: %d", err);
+		}
+	}
+}
+
+static void connected_entry(void *o)
+{
+	LOG_INF("MQTT connected");
+
+	ARG_UNUSED(o);
+
+	int err;
+	struct mqtt_topic topics[] = {
+		{
+			.topic.utf8 = CONFIG_MQTT_SAMPLE_TRANSPORT_SUBSCRIBE_TOPIC,
+			.topic.size = strlen(CONFIG_MQTT_SAMPLE_TRANSPORT_SUBSCRIBE_TOPIC),
+		},
+	};
+	struct mqtt_subscription_list list = {
+		.list = topics,
+		.list_count = ARRAY_SIZE(topics),
+		.message_id = k_uptime_get_32(),
+	};
+
+	err = mqtt_helper_subscribe(&list);
+	if (err) {
+		LOG_ERR("Failed to subscribe to topics, error: %d", err);
+		return;
+	}
+}
+
+static void connected_run(void *o)
+{
+	int err;
+	struct s_object *od = o;
+
+	if (od->status == NETWORK_DISCONNECTED) {
```

Comment:
Is this safe? status is in a union with payload. You do not know which channel that triggered the run of this function. And status would then be a byte in byte in payload which by accident could be NETWORK_DISCONNECTED (0x00)